### PR TITLE
Back pressure fix for 3.7

### PR DESCRIPTION
**Pull request by bentoi**
Here's a proposal for fixing back pressure on 3.7. The connection/thread pool implementation is now very close to the C++/IOCP implementation. The main difference is that IO operations are now started on the .NET thread pool to ensure it won't be interrupted if the thread exits. I'm undecided on whether or not it's a good idea to commit this large fix. We can use it for 3.8 however.

### CODE REVIEW

File: csharp/test/Ice/ami/AllTests.cs [hunk 1]
```diff
@@ -3842,6 +3842,33 @@ public static void allTests(global::Test.TestHelper helper, bool collocated)
...
+                output.Flush();
+                {
+                    // Keep the 3 server thread pool threads busy.
+                    Task sleep1Task = p.sleepAsync(1000);
+                    Task sleep2Task = p.sleepAsync(1000);
+                    Task sleep3Task = p.sleepAsync(1000);
+                    bool canceled = false;
+                    try
+                    {
+                        var cts = new CancellationTokenSource(200);
```
Comment (bernardnormier) [inline csharp/test/Ice/ami/AllTests.cs:3855]: using var cts

File: csharp/test/Ice/ami/AllTests.cs [hunk 2]
```diff
@@ -3842,6 +3842,33 @@ public static void allTests(global::Test.TestHelper helper, bool collocated)
...
+                        var onewayProxy = (Test.TestIntfPrx)p.ice_oneway();
+
+                        // Sending should be canceled because the TCP receive buffer size on the server is set to 50KB
+                        onewayProxy.opWithPayloadAsync(new byte[768 * 1024]).Wait(cts.Token);
```
Comment (bernardnormier) [inline csharp/test/Ice/ami/AllTests.cs:3859]: That's because we don't use async/await anywhere in 3.7?

We should fix this code when porting it to 3.8.

**Review by bernardnormier — COMMENTED** [review]
> I'm undecided on whether or not it's a good idea to commit this large fix. 

I think we should fix it on 3.7, provided it doesn't change the external behavior (other than back pressure).

As I understand it:
 - we read (and write?) from a .NET threal pool thread with Task.Run() in startAsync
 - upon completion, we queue a work item in the Ice thread pool

But then, how do we ensure we read only when the Ice thread pool is not exhausted? Is it because startAsync is called by the work item we queued?

**Comment by bentoi** [comment]
> As I understand it:
> 
> * we read (and write?) from a .NET threal pool thread with Task.Run() in startAsync

Reads and writes are _started_ on the .NET thread pool and when completed, their completion callbacks are also executed on the .NET thread pool.

> * upon completion, we queue a work item in the Ice thread pool

Right.

> But then, how do we ensure we read only when the Ice thread pool is not exhausted? Is it because startAsync is called by the work item we queued?

If upon read completion the Ice thread pool is exhausted, the Ice thread pool queued work item won't execute until there's a free thread. So reads and writes won't progress.

**Comment by bentoi** [comment]
I've push additional fixes and add C++/Java tests.

File: cpp/test/Ice/ami/AllTests.cpp [hunk 3]
```diff
@@ -2482,6 +2485,28 @@ allTests(Test::TestHelper* helper, bool collocated)
...
+        cout << "testing back pressure... " << flush;
+        {
+            // Keep the 3 server thread pool threads busy.
+            auto sleep1Future = p->sleepAsync(1000);
+            auto sleep2Future = p->sleepAsync(1000);
+            auto sleep3Future = p->sleepAsync(1000);
+
+            auto onewayProxy = Ice::uncheckedCast<Test::TestIntfPrx>(p->ice_oneway());
+
+            // Sending should be canceled because the TCP receive buffer size on the server is set to
```
Comment (bernardnormier) [inline cpp/test/Ice/ami/AllTests.cpp:2497]: "canceled" doesn't sound appropriate here. I would use "block" or "not complete".

File: cpp/test/Ice/ami/AllTests.cpp [hunk 4]
```diff
@@ -2482,6 +2485,28 @@ allTests(Test::TestHelper* helper, bool collocated)
...
+            // 50KB
```
Comment (bernardnormier) [inline cpp/test/Ice/ami/AllTests.cpp:2498]: There is no client-side send buffer? Only the size of the server side buffer matters?

File: csharp/file.wav
Comment (bernardnormier) [inline csharp/file.wav:1]: What is this file?

File: csharp/test/Ice/ami/AllTests.cs [hunk 5]
```diff
@@ -3842,6 +3842,36 @@ public static void allTests(global::Test.TestHelper helper, bool collocated)
...
+                    Task sleep2Task = p.sleepAsync(1000);
+                    Task sleep3Task = p.sleepAsync(1000);
+                    bool canceled = false;
+                    using(var cts = new CancellationTokenSource(200))
+                    {
+                        try
+                        {
+                            var onewayProxy = (Test.TestIntfPrx)p.ice_oneway();
+
+                            // Sending should be canceled because the TCP receive buffer size on the server is set to
```
Comment (bernardnormier) [inline csharp/test/Ice/ami/AllTests.cs:3859]: "canceled" is ok here since the call is indeed canceled.

File: csharp/test/Ice/ami/AllTests.cs [hunk 6]
```diff
@@ -3842,6 +3842,36 @@ public static void allTests(global::Test.TestHelper helper, bool collocated)
...
+                            // 50KB
+                            onewayProxy.opWithPayloadAsync(new byte[768 * 1024]).Wait(cts.Token);
```
Comment (bernardnormier) [inline csharp/test/Ice/ami/AllTests.cs:3861]: Why not use the cancel parameter of opWithPayloadAsync?

https://doc.zeroc.com/ice/3.7/language-mappings/c-sharp-mapping/client-side-slice-to-c-sharp-mapping/asynchronous-method-invocation-ami-in-c-sharp/ami-in-c-sharp-with-tasks#id-.AMIinCSharpwithTasksv3.7-CancelingAsynchronousRequestsinC#

**Review by bernardnormier — APPROVED** [review]
Looks good.

File: cpp/test/Ice/ami/AllTests.cpp (hunk 4 above)
Comment (bentoi) [reply cpp/test/Ice/ami/AllTests.cpp:2498]: Yes, we also set the send buffer size, fixed the comment.

File: csharp/file.wav
Comment (bentoi) [reply csharp/file.wav:1]: No idea how this ended up here. Removed.

File: csharp/test/Ice/ami/AllTests.cs (hunk 6 above)
Comment (bentoi) [reply csharp/test/Ice/ami/AllTests.cs:3861]: I tried but that's another weirdness of Ice for C#: the cancellation of an operation with the cancellation token doesn't follow the C# pattern. It works like the invocation timeout and like C++/Java it raises `InvocationCanceledException`. An operation which isn't being sent is not canceled until the payload is sent.

**Pull request merged**